hal/halx86/legacy/halpnpdd.c

```diff
@@ -1,931 +1,544 @@
-        {
-            /* Count how many PDOs we have */
-            PdoExtension = FdoExtension->ChildPdoList;
-            while (PdoExtension)
-            {
-                /* Next one */
-                PdoExtension = PdoExtension->Next;
-                PdoCount++;
-            }
-
-            /* Add the PDOs that already exist in the device relations */
-            if (*DeviceRelations)
-            {
-                PdoCount += (*DeviceRelations)->Count;
-            }
-
-            /* Allocate our structure */
-            FdoRelations = ExAllocatePoolWithTag(PagedPool,
-                                                 FIELD_OFFSET(DEVICE_RELATIONS,
-                                                              Objects) +
-                                                 sizeof(PDEVICE_OBJECT) * PdoCount,
-                                                 TAG_HAL);
-            if (!FdoRelations) return STATUS_INSUFFICIENT_RESOURCES;
-
-            /* Save our count */
-            FdoRelations->Count = PdoCount;
-
-            /* Query existing relations */
-            ObjectEntry = FdoRelations->Objects;
-            if (*DeviceRelations)
-            {
-                /* Check if there were any */
-                if ((*DeviceRelations)->Count)
-                {
-                    /* Loop them all */
-                    do
-                    {
-                        /* Copy into our structure */
-                        *ObjectEntry++ = (*DeviceRelations)->Objects[i];
-                    }
-                    while (++i < (*DeviceRelations)->Count);
-                }
-
-                /* Free existing structure */
-                ExFreePool(*DeviceRelations);
-            }
-
-            /* Now check if we have a PDO list */
-            PdoExtension = FdoExtension->ChildPdoList;
-            if (PdoExtension)
-            {
-                /* Loop the PDOs */
-                do
-                {
-                    /* Save our own PDO and reference it */
-                    *ObjectEntry++ = PdoExtension->PhysicalDeviceObject;
-                    ObReferenceObject(PdoExtension->PhysicalDeviceObject);
-
-                    /* Go to our next PDO */
-                    PdoExtension = PdoExtension->Next;
-                }
-                while (PdoExtension);
-            }
-
-            /* Return the new structure */
-            *DeviceRelations = FdoRelations;
-            return STATUS_SUCCESS;
-        }
-    }
-    else
-    {
-        /* The only other thing we support is a target relation for the PDO */
-        if ((RelationType == TargetDeviceRelation) &&
-            (ExtensionType == PdoExtensionType))
+        PFDO_EXTENSION fdoExtension = DeviceObject->DeviceExtension;
+        PDEVICE_RELATIONS fdoRelations = ExAllocatePoolWithTag(
+            PagedPool,
+            FIELD_OFFSET(DEVICE_RELATIONS, Objects)
+                + (sizeof(PDEVICE_OBJECT) * fdoExtension->PdoCount),
```

Comment:
Maybe something like this might be more suitable:
```suggestion
            FIELD_OFFSET(DEVICE_RELATIONS, Objects[fdoExtension->PdoCount]),
```

Reply:
With MSVC sometimes this syntax works, and sometimes it doesn't (but I don't remember exactly in which condition).